dbt/adapters/athena/impl.py

```diff
@@ -216,33 +218,36 @@ def apply_lf_grants(self, relation: AthenaRelation, lf_grants_config: Dict[str,
 
     @available
     def is_work_group_output_location_enforced(self) -> bool:
-        conn = self.connections.get_thread_connection()
-        creds = conn.credentials
-        client = conn.handle
+        if self.work_group_output_location_enforced is None:
+            conn = self.connections.get_thread_connection()
+            creds = conn.credentials
+            client = conn.handle
 
-        with boto3_client_lock:
-            athena_client = client.session.client(
-                "athena",
-                region_name=client.region_name,
-                config=get_boto3_config(num_retries=creds.effective_num_retries),
-            )
+            with boto3_client_lock:
```

Comment:
let's do another if after lock acquisition, if this flag is already calculated then just return

Reply:
please read more on the multi-threading / concurrent executions and how locks / semaphores works. I still think that there is can be a situation that after released lock it will appear that flag is already not none